pyard/data_repository.py

```diff
@@ -304,7 +318,37 @@ def generate_alleles_and_xx_codes_and_who(db_connection: sqlite3.Connection, img
     db.save_dict(db_connection, 'who_group', flat_who_group,
                  columns=('who', 'allele_list'))
 
-    return valid_alleles, who_alleles, xx_codes, who_group
+    # shortnulls
+    # scan WHO alleles for those with expression characters and make shortnull mappings 
+    # DRB4*01:03N | DRB4*01:03:01:02N/DRB4*01:03:01:13N 
+    # DRB5*01:08N | DRB5*01:08:01N/DRB5*01:08:02N 
+    shortnulls = dict()
+    for k in who_group:
+        # e.g. DRB4*01:03
+        ea=[]
+        expression_chars_found = dict()
+        for al in who_group[k]:
+            # if an allele in a who_group has an expression character but the group allele doesnt, 
+            # add it to shortnulls
+            if al[-1] in expression_chars and k[-1] not in expression_chars and k[-1] not in ['G', 'P'] and ":" in k:
+                # e.g. DRB4*01:03:01:02N
+                ex = al[-1]
+                if ex not in expression_chars_found:
```

Comment:
Use `set` instead to keep unique values.
```
expression_chars_found = set()
...
expression_chars_found.add(ex)
...
if expression_chars_found:
  ...add to shortnull...

Reply:
yes!